require-returns-check returning false errors with loops

This is similar to #924 but doesn't seem to be fixed.

## Expected behavior
<!-- Provide a detailed description of how you expected the software to -->
<!-- behave. -->

There should be no `jsdoc/require-returns-check` errors reported for the functions below. This was working for eslint-plugin-jsdoc v39.3.6.

```js
/**
 * @return {number}
 */
function foo() {
  while (true) {
    const n = Math.random();
    if (n < 0.5) {
      return n;
    }
  }
}
```

## Actual behavior
<!-- Provide a detailed description of how the software actually behaved, -->
<!-- including any rationale for why that behavior is incorrect. -->

```
  58:1  error  JSDoc @return declaration present but return expression not available in function  jsdoc/require-returns-check
```

## Environment

- Node version: v16.13.0
- ESLint version: 8.26.0
- `eslint-plugin-jsdoc` version: 39.6.2

--- a/README.md
+++ b/README.md
@@ -17881,6 +17881,20 @@ function foo() {
     }
 };
 // Message: JSDoc @returns declaration present but return expression not available in function.
+
+/**
+ * @returns {number}
+ */
+function foo() {
+  let n = 1;
+  while (n > 0.5) {
+    n = Math.random();
+    if (n < 0.2) {
+      return n;
+    }
+  }
+}
+// Message: JSDoc @returns declaration present but return expression not available in function.
 ````
 
 The following patterns are not considered problems:
@@ -18517,6 +18531,18 @@ const quux = (someVar) => {
     return true;
   }
 };
+
+/**
+ * @returns {number}
+ */
+function foo() {
+  while (true) {
+    const n = Math.random();
+    if (n < 0.5) {
+      return n;
+    }
+  }
+}
 ````
 
 

--- a/src/utils/hasReturnValue.js
+++ b/src/utils/hasReturnValue.js
@@ -149,9 +149,16 @@ const allBrancheshaveReturnValues = (node, promFilter) => {
     return allBrancheshaveReturnValues(lastBodyNode, promFilter);
   }
 
-  case 'LabeledStatement':
   case 'WhileStatement':
   case 'DoWhileStatement':
+    if (node.test.value === true) {
+      // If this is an infinite loop, we assume only one branch
+      //   is needed to provide a return
+      return hasReturnValue(node.body, false, promFilter);
+    }
+
+    // Fallthrough
+  case 'LabeledStatement':
   case 'ForStatement':
   case 'ForInStatement':
   case 'ForOfStatement':

--- a/test/rules/assertions/requireReturnsCheck.js
+++ b/test/rules/assertions/requireReturnsCheck.js
@@ -654,6 +654,29 @@ export default {
         },
       ],
     },
+    {
+      code: `
+        /**
+         * @returns {number}
+         */
+        function foo() {
+          let n = 1;
+          while (n > 0.5) {
+            n = Math.random();
+            if (n < 0.2) {
+              return n;
+            }
+          }
+        }
+      `,
+      errors: [
+        {
+          line: 2,
+          message: 'JSDoc @returns declaration present but return expression not available in function.',
+        },
+      ],
+    },
+
   ],
   valid: [
     {
@@ -1575,6 +1598,21 @@ export default {
         }
       };
   `,
+    },
+    {
+      code: `
+        /**
+         * @returns {number}
+         */
+        function foo() {
+          while (true) {
+            const n = Math.random();
+            if (n < 0.5) {
+              return n;
+            }
+          }
+        }
+      `,
     },
   ],
 };

